apps/argus/src/services/controller_service.rs

```diff
@@ -0,0 +1,126 @@
+//! Controller Service
+//!
+//! This service orchestrates the price update process for a given blockchain network.
+//! It reads from the SubscriptionState, PythPriceState, and ChainPriceState to determine
+//! whether to update the on-chain price for a given subscription. It also triggers the
+//! PricePusherService to push the update to the target blockchain network.
+
+use anyhow::Result;
+use async_trait::async_trait;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::watch;
+use tokio::time;
+use tracing;
+
+use crate::adapters::types::{PriceId, SubscriptionId};
+use crate::services::types::PushRequest;
+use crate::services::Service;
+use crate::state::ChainName;
+use crate::state::{ChainPriceState, PythPriceState, SubscriptionState};
+
+pub struct ControllerService {
+    name: String,
+    update_interval: Duration,
+    subscription_state: Arc<SubscriptionState>,
+    pyth_price_state: Arc<PythPriceState>,
+    chain_price_state: Arc<ChainPriceState>,
+}
+
+impl ControllerService {
+    pub fn new(
+        chain_name: ChainName,
+        update_interval: Duration,
+        subscription_state: Arc<SubscriptionState>,
+        pyth_price_state: Arc<PythPriceState>,
+        chain_price_state: Arc<ChainPriceState>,
```

Comment:
ah so instead of relying on the magic of type inferences you decided to literally pass the states around. good idea!